coreneuron/gpu/nrn_acc_manager.cpp

```diff
@@ -400,6 +400,46 @@ void setup_nrnthreads_on_device(NrnThread* threads, int nthreads) {
         } else {
             printf("\n WARNING: NrnThread %d not permuted, error for linear algebra?", i);
         }
+
+        {
+            TrajectoryRequests* tr = nt->trajec_requests;
+            if (tr) {
+                // Create a device-side copy of the `trajec_requests` struct and
+                // make sure the device-side NrnThread object knows about it.
+                auto* d_trajec_requests = reinterpret_cast<TrajectoryRequests*>(
+                    acc_copyin(tr, sizeof(TrajectoryRequests)));
+                acc_memcpy_to_device(&(d_nt->trajec_requests),
+                                     &d_trajec_requests,
+                                     sizeof(TrajectoryRequests*));
+                // Initialise the double** varrays member of the struct.
+                auto* d_tr_varrays = reinterpret_cast<double**>(
+                    acc_copyin(tr->varrays, sizeof(double*) * tr->n_trajec));
+                acc_memcpy_to_device(&(d_trajec_requests->varrays),
+                                     &d_tr_varrays,
+                                     sizeof(double**));
+                // Initialise the double** gather member of the struct.
+                auto* d_tr_gather = reinterpret_cast<double**>(
+                    acc_copyin(tr->gather, sizeof(double*) * tr->n_trajec));
+                acc_memcpy_to_device(&(d_trajec_requests->gather), &d_tr_gather, sizeof(double**));
+                // `varrays` and `gather` are minimal, the host-size buffers have
+                // `bsize` elements.
+                for (int i = 0; i < tr->n_trajec; ++i) {
+                    // tr->varrays[i] is a buffer of tr->bsize doubles on the host,
+                    // make a device-side copy of it and store a pointer to it in
+                    // the device-side version of tr->varrays.
+                    auto* d_buf_traj_i = reinterpret_cast<double*>(
+                        acc_copyin(tr->varrays[i], tr->bsize * sizeof(double)));
+                    acc_memcpy_to_device(&(d_tr_varrays[i]), &d_buf_traj_i, sizeof(double*));
+                    // tr->gather[i] is a double* referring to (host) data in the
+                    // (host) _data block
+                    auto* d_gather_i = acc_deviceptr(tr->gather[i]);
+                    acc_memcpy_to_device(&(d_tr_gather[i]), &d_gather_i, sizeof(double*));
+                }
+                // TODO: other members of the TrajectoryRequests struct:
+                // `scatter`, `n_pr`, `n_trajec`, `bsize`, `vsize` are not
+                // updated on device.
```

Comment:
scalar members are implicitly updated by the copy?
```suggestion
                // TODO: `scatter` member of the TrajectoryRequests struct
                // is not updated on device.
```

Reply:
Yes..it was a bad comment. There is also a `void** vpr` that was not mentioned in the comment.

For the scalar members: you're right these are automatically updated when we call `acc_copyin`. That said, `vsize` changes during the simulation and is not kept up to date on the device.

I'll update the comment...